"Exception thrown during asynchronous load" when using quarkus-cache and quarkus-hibernate-validator

**Describe the bug**
When using service validation on a method and also use caching on that service method an exception is written to the log. 

**Expected behavior**
No exception is written to the log.

**Actual behavior**
Something like this exception is written to the log:

```
WARNING: Exception thrown during asynchronous load
java.util.concurrent.CompletionException: javax.validation.ConstraintViolationException: 1 constraint violation(s) occurred during method validation.
Constructor or Method: public java.lang.String ...Service.doService(ServiceRequest)
Argument values: [...ServiceRequest@7a0f43dc]
Constraint violations: 
 (1) Kind: PROPERTY
 message: darf nicht null sein
 root bean: ...Service_Subclass@15ec2ac4
 property path: doService.arg0.text
 constraint: @javax.validation.constraints.NotNull(message="{javax.validation.constraints.NotNull.message}", payload={}, groups={})
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)
	at io.smallrye.context.SmallRyeThreadContext$ContextualRunnable.run(SmallRyeThreadContext.java:56)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)

	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: javax.validation.ConstraintViolationException: 1 constraint violation(s) occurred during method validation.
Constructor or Method: public java.lang.String Service.doService(...ServiceRequest)
Argument values: [...ServiceRequest@7a0f43dc]
Constraint violations: 
 (1) Kind: PROPERTY
 message: darf nicht null sein
 root bean: Service_Subclass@15ec2ac4
 property path: service.arg0.text
 constraint: @javax.validation.constraints.NotNull(message="{javax.validation.constraints.NotNull.message}", payload={}, groups={})
	at io.quarkus.hibernate.validator.runtime.interceptor.AbstractMethodValidationInterceptor.validateMethodInvocation(AbstractMethodValidationInterceptor.java:65)
	at io.quarkus.hibernate.validator.runtime.interceptor.MethodValidationInterceptor.validateMethodInvocation(MethodValidationInterceptor.java:17)
	at io.quarkus.hibernate.validator.runtime.interceptor.MethodValidationInterceptor_Bean.intercept(MethodValidationInterceptor_Bean.zig:340)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.quarkus.cache.runtime.CacheResultInterceptor.lambda$intercept$0(CacheResultInterceptor.java:27)
	at io.quarkus.cache.runtime.caffeine.CaffeineCache$MappingSupplier.get(CaffeineCache.java:136)
	at io.quarkus.cache.runtime.caffeine.CaffeineCache.lambda$get$0(CaffeineCache.java:58)
	at com.github.benmanes.caffeine.cache.LocalAsyncCache.lambda$get$0(LocalAsyncCache.java:77)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	... 9 more
```



**To Reproduce**
In the pom.xml add:
```
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-cache</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-hibernate-validator</artifactId>
    </dependency>
```

Create a service like this:
```
@ApplicationScoped
public class Service {
  @CacheResult(cacheName = "cache")
  public String doService(@Valid ServiceRequest request) {
    // do something important
  }
}
```

Create a ServiceRequest POJO with validation annotations:
```
public class ServiceRequest{
  @NotNull
  private String text;
  // getters / setters
}
```

Invoke the service method from a QuarkusTest or REST service (with text = null), then the error will appear. It doesn't if I remove the cache annotation.

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Microsoft Windows [Version 10.0.17763.1282]
 - Output of `java -version`: java version "11.0.7" 2020-04-14 LTS
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.6.1.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)

--- a/extensions/cache/deployment/src/test/java/io/quarkus/cache/test/runtime/ThrowExecutionExceptionCauseTest.java
+++ b/extensions/cache/deployment/src/test/java/io/quarkus/cache/test/runtime/ThrowExecutionExceptionCauseTest.java
@@ -1,0 +1,61 @@
+package io.quarkus.cache.test.runtime;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import java.io.IOException;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.cache.CacheResult;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ThrowExecutionExceptionCauseTest {
+
+    private static final String FORCED_EXCEPTION_MESSAGE = "Forced exception";
+
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClass(CachedService.class));
+
+    @Inject
+    CachedService cachedService;
+
+    @Test
+    public void testRuntimeExceptionThrow() {
+        NumberFormatException e = assertThrows(NumberFormatException.class, () -> {
+            cachedService.runtimeExceptionCachedMethod();
+        });
+        assertEquals(FORCED_EXCEPTION_MESSAGE, e.getMessage());
+    }
+
+    @Test
+    public void testCheckedExceptionThrow() {
+        IOException e = assertThrows(IOException.class, () -> {
+            cachedService.checkedExceptionCachedMethod();
+        });
+        assertEquals(FORCED_EXCEPTION_MESSAGE, e.getMessage());
+    }
+
+    @ApplicationScoped
+    static class CachedService {
+
+        private static final String CACHE_NAME = "test-cache";
+
+        @CacheResult(cacheName = CACHE_NAME)
+        public String runtimeExceptionCachedMethod() {
+            throw new NumberFormatException(FORCED_EXCEPTION_MESSAGE);
+        }
+
+        @CacheResult(cacheName = CACHE_NAME)
+        public String checkedExceptionCachedMethod() throws Exception {
+            throw new IOException(FORCED_EXCEPTION_MESSAGE);
+        }
+    }
+}

--- a/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheException.java
+++ b/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheException.java
@@ -1,0 +1,10 @@
+package io.quarkus.cache.runtime;
+
+public class CacheException extends RuntimeException {
+
+    private static final long serialVersionUID = -3465350968605912384L;
+
+    public CacheException(Throwable cause) {
+        super(cause);
+    }
+}

--- a/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheResultInterceptor.java
+++ b/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheResultInterceptor.java
@@ -1,6 +1,7 @@
 package io.quarkus.cache.runtime;
 
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executor;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
@@ -41,7 +42,11 @@ public class CacheResultInterceptor extends CacheInterceptor {
                             return getValueLoader(context, executor);
                         }
                     });
-            return cacheValue.get();
+            try {
+                return cacheValue.get();
+            } catch (ExecutionException e) {
+                throw getExceptionToThrow(e);
+            }
         } else {
 
             // The lock timeout logic starts here.
@@ -66,12 +71,18 @@ public class CacheResultInterceptor extends CacheInterceptor {
             if (isCurrentThreadComputation[0]) {
                 // The value is missing and its computation was started from the current thread.
                 // We'll wait for the result no matter how long it takes.
-                return cacheValue.get();
+                try {
+                    return cacheValue.get();
+                } catch (ExecutionException e) {
+                    throw getExceptionToThrow(e);
+                }
             } else {
                 // The value is either already present in the cache or missing and its computation was started from another thread.
                 // We want to retrieve it from the cache within the lock timeout delay.
                 try {
                     return cacheValue.get(binding.lockTimeout(), TimeUnit.MILLISECONDS);
+                } catch (ExecutionException e) {
+                    throw getExceptionToThrow(e);
                 } catch (TimeoutException e) {
                     // Timeout triggered! We don't want to wait any longer for the value computation and we'll simply invoke the
                     // cached method and return its result without caching it.
@@ -88,12 +99,25 @@ public class CacheResultInterceptor extends CacheInterceptor {
             public Object get() {
                 try {
                     return context.proceed();
-                } catch (RuntimeException e) {
-                    throw e;
                 } catch (Exception e) {
-                    throw new RuntimeException(e);
+                    throw new CacheException(e);
                 }
             }
         }, executor);
+    }
+
+    private Exception getExceptionToThrow(ExecutionException e) {
+        if (e.getCause() instanceof CacheException && e.getCause().getCause() instanceof Exception) {
+            return (Exception) e.getCause().getCause();
+        } else {
+            /*
+             * If:
+             * - the cause is not a CacheException
+             * - the cause is a CacheException which doesn't have a cause itself
+             * - the cause is a CacheException which was caused itself by an Error
+             * ... then we'll throw the original ExecutionException.
+             */
+            return e;
+        }
     }
 }

